python/lsst/meas/algorithms/stamps.py

```diff
@@ -0,0 +1,209 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+"""Collection of small images (stamps), each centered on a bright star.
+"""
+
+__all__ = ["Stamp", "Stamps"]
+
+import collections.abc
+from dataclasses import dataclass
+import numpy
+
+import lsst.afw.image as afwImage
+import lsst.afw.fits as afwFits
+from lsst.geom import Box2I, Point2I, Extent2I, Angle
+from lsst.daf.base import PropertySet
+
+
+class Stamp(dataclass):
+    """Single stamp
+    """
+    stamp: afwImage.maskedImage.MaskedImageF
+    ra: Angle
+    dec: Angle
+    size: int
+
+    @classmethod
+    def factory(cls, stamp, metadata, index):
+        if 'RA_DEG' in metadata.keys() and 'DEC_DEG' in metadata.keys() and 'SIZE' in metadata.keys():
+            return cls(stamp, Angle(metadata['RA_DEG'][index]), Angle(metadata['DEC_DEG'][index]),
+                       metadata['SIZE'][index])
+        else:
+            return cls(stamp=stamp, ra=Angle(numpy.nan), dec=Angle(numpy.nan), size=-1)
+
+
+class Stamps(collections.abc.Sequence):
+    """Collection of  stamps and associated metadata.
+
+    Parameters
+    ----------
+    stamps : `collections.abc.Sequence` [`Stamp`]
+        Sequence of stamps.
+    metadata : `lsst.daf.base.PropertyList`, optional
+        Metadata associated with the bright stars.
+
+    Raises
+    ------
+    ValueError
+        Raises if the stamps are not the correct type
+
+    Notes
+    -----
+    A (gen2) butler can be used to read only a part of the stamps,
+    specified by a bbox:
+
+    >>> starSubregions = butler.get("brightStarStamps_sub", dataId, bbox=bbox)
+    """
+
+    def __init__(self, stamps, metadata=None, has_mask=True, has_variance=True):
+        for item in stamps:
+            if not isinstance(item, Stamp):
+                raise ValueError(f"Can only add instances of Stamp, got {type(item)}")
+        self._stamps = stamps
+        self._metadata = PropertySet() if metadata is None else metadata.deepCopy()
+        self._metadata['HAS_MASK'] = has_variance
+        self._metadata['HAS_VARIANCE'] = has_variance
+
+    def __len__(self):
+        return len(self._stamps)
+
+    def __getitem__(self, index):
+        return self._stamps[index]
+
+    def __iter__(self):
+        return iter(self._stamps)
+
+    def _refresh_metadata(self):
+        # ensure metadata contains current number of objects
+        self._metadata["N_STAMPS"] = len(self)
```

Comment:
I wouldn't think that this method should be necessary at all, if `N_STAMPS` is generated only when necessary.